DOCS-1524: Add PID config to motors and bases (#2547)
viamrobotics/docs: static/include/components/motor-sensor.md

If you want to control your motor by specifying the velocity in terms of distance/time, for example 5 m/s and you have a sensor, for example a wheel encoder, that measures the actual value, you can make use of the sensor to adjust the velocity.

One way to adjust velocity based on a sensor's measurements, is using a PID control loop that measures the velocity of movement and determines the difference or error from the desired velocity to compute a correction for the motor based on three terms:

- A _proportional_ term that is the current error
- An _integral_ term that is the total cumulative error
- A _derivative_ term that is the rate of change of the error

By tuning the coefficients on each of these terms, you can adjust how your motor converges towards the target value, how quickly the system reaches the target value, and how much the system overshoots when approaching the target value.

> If you use `{ "p": 0, "i": 0, "d": 0 }`, `viam-server` auto-tunes the values and logs the tuned values. To avoid retuning the control parameters constantly, copy the values from the log and add them to the configuration once tuned.

Author: ```suggestion
If you use `{ "p": 0, "i": 0, "d": 0 }`, `viam-server` auto-tunes the values and logs the tuned values.
Tuning takes several seconds and spins the motors.
To avoid tuning every time the robot starts up, copy the values from the log and add them to the configuration once tuned.
```

> If you want to control your motor by specifying the velocity in terms of distance/time, for example 5 m/s and you have a sensor, for example a wheel encoder, that measures the actual value, you can make use of the sensor to adjust the velocity.

Reviewer: ```suggestion
If you want to control your motor by specifying the distance and velocity in terms of distance and distance/time, for example 2 rotations at 5 m/s and you have an encoder, you can make use of the position of your encoder to adjust the distance and velocity.
```

> One way to adjust velocity based on a sensor's measurements, is using a PID control loop that measures the velocity of movement and determines the difference or error from the desired velocity to compute a correction for the motor based on three terms:

Reviewer: ```suggestion
One way to adjust distance and velocity based on an encoder's measurements, is using a PID control loop that measures the current position and change in position and determines the difference or error from the desired distance and velocity to compute a correction for the motor based on three terms:
```

> One way to adjust distance and velocity based on an encoder's measurements, is using a PID control loop that measures the current position and change in position and determines the difference or error from the desired distance and velocity to compute a correction for the motor based on three terms:

Reviewer 2: ```suggestion
The 'gpio' model of Motor for Encoded Motors uses a PID control loop that measures the current position and change in position and determines the difference or error from the desired distance and velocity to compute a correction for the motor based on three terms:
```

Reviewer 2: @martha-johnston , confirming that for encoded gpio motors, PID control is now always used (and the "10% increment" form of control is never used)?

Reviewer: no, if you don't add the control_parameters the old version is used (the 10% thing)

Reviewer 2: Ah right, sorry I got confused between gpio motor and Spin 🤦‍♀️ 

Makes sense, updating the suggested edit above accordingly

Reviewer 2: updated

- A _proportional_ term that is the current error
- An _integral_ term that is the total cumulative error
- A _derivative_ term that is the rate of change of the error

> By tuning the coefficients on each of these terms, you can adjust how your motor converges towards the target value, how quickly the system reaches the target value, and how much the system overshoots when approaching the target value.

Reviewer: ```suggestion
By tuning the coefficients on each of these terms, you can adjust how your motor converges towards the target values, how quickly the system reaches the target values, and how much the system overshoots when approaching the target values.
```

> If you use `{ "p": 0, "i": 0, "d": 0 }`, `viam-server` auto-tunes the values and logs the tuned values.

Reviewer: ```suggestion
If you want these values to be auto-tuned, you can set all values to 0: `{ "p": 0, "i": 0, "d": 0 }`, and `viam-server` will auto-tune and log the calculated values.
```

> Tuning takes several seconds and spins the motors.

Reviewer: ```suggestion
Tuning takes several seconds and spins the motor.
```

> To avoid tuning every time the robot starts up, copy the values from the log and add them to the configuration once tuned.

Reviewer: ```suggestion
To avoid tuning every time the robot starts up, copy the values from the logs and add them to the configuration once tuned.
```

> To avoid tuning every time the robot starts up, copy the values from the logs and add them to the configuration once tuned.

Reviewer 2: [nit/optional] This is also mentioned in the attributes table, so I think it'd be okay to leave it out of this section

> If you want to control your motor by specifying the distance and velocity in terms of distance and distance/time, for example 2 rotations at 5 m/s and you have an encoder, you can make use of the position of your encoder to adjust the distance and velocity.

Reviewer 2: ```suggestion
If you want to control your motor by specifying the distance and velocity in terms of distance and distance/time, for example 2 rotations at 5 m/s, and you have an encoder, you can make use of the position of your encoder to adjust the distance and velocity.
```

You can do this in two ways.

The first just uses the movement sensor feedback to increase or decrease the power being set on the motors in increments of 10%.

The second requires setting the `control_parameters` attribute and uses a PID control loop that measures the current position and change in position and determines the difference or error from the desired distance and velocity to compute a correction for the motor based on three terms:

- A _proportional_ term that is the current error
- An _integral_ term that is the total cumulative error
- A _derivative_ term that is the rate of change of the error

By tuning the coefficients on each of these terms, you can adjust how your motor converges towards the target values, how quickly the system reaches the target values, and how much the system overshoots when approaching the target values.

If you want these values to be auto-tuned, you can set all values to 0: `{ "p": 0, "i": 0, "d": 0 }`, and `viam-server` will auto-tune and log the calculated values.
Tuning takes several seconds and spins the motor.

> To avoid tuning every time the robot starts up, copy the values from the logs and add them to the configuration once tuned.

Author: ```suggestion
Copy the values from the logs and add them to the configuration once tuned for the values to take effect.
```
